torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3244,39 +3404,62 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
-                if user_tb:
-                    maybe_user_loc = " at " + format_frame(user_tb[-1])
-
-                is_debug = self.log.isEnabledFor(logging.DEBUG)
-                maybe_extra_debug = ""
-                if is_debug and user_tb:
-                    maybe_extra_debug = (
-                        '\nUser Stack (most recent call last):\n' +
-                        '  (snipped, see stack below for prefix)\n' +
-                        ''.join(traceback.format_list(user_tb))
-                    )
-                self.log.info(
-                    "eval %s [guard added]%s (%s)%s",
-                    g,
-                    maybe_user_loc,
-                    format_frame(frame),
-                    maybe_extra_debug,
-                    stack_info=is_debug,
-                )
+            self._log_guard("eval", g, tb)
         else:
             self.log.debug("eval %s [guard suppressed]", g)
 
         return concrete_val
 
+    def defer_runtime_assert(self, orig_expr: "sympy.Expr", msg, fx_node=None):
+        expr = orig_expr
+
+        static_expr = self._maybe_evaluate_static(expr)
+        if static_expr is not None:
+            self.log.debug("runtime_assert %s == %s [statically known]", orig_expr, static_expr)
+            return static_expr
+
+        # Attempt to eliminate the unbacked SymInt
+        new_expr = self._maybe_evaluate_static(expr, unbacked_only=True)
+        if new_expr.free_symbols <= self.var_to_val.keys():
+            # Do a normal guard
+            return self.evaluate_expr(new_expr, fx_node=fx_node)
+        # NB: Don't use new_expr as expr; it could contain gunk like shape0
+        # which we don't want to guard on
+
+        # OK, we're definitely doing a runtime assert now
+        if (
+            _translation_validation_enabled()
+            and fx_node is not None
+            and not self._suppress_guards_tls()
+        ):
+            self.create_fx_call_function(torch._assert, (fx_node,))
+
+        self._check_frozen(expr, sympy.true)
+
+        # TODO: eliminate symbols on equality tests
+        # (_maybe_guard_eq assumes everything is hinted so it doesn't work
+        # here)
+
+        if not self._suppress_guards_tls():
+            tb = traceback.extract_stack()[:-1]
+            stack = ''.join(traceback.format_list(tb))
+            ra = RuntimeAssert(expr, msg, stack)
+            # TODO: Do this in a way that is less janky than int(s.name[1:])
+            cands = sorted([s for s in expr.free_symbols if s.name.startswith("i")], key=lambda s: int(s.name[1:]))
+            self.deferred_runtime_asserts[cands[-1]].append(ra)
+            self.num_deferred_runtime_asserts += 1
+            # TODO: refine ranges
+            # Unfortunately, range refinement is probably going to not
+            # work most of the time, because we don't support symbols
+            # in ranges.  For example, i0 <= s0 is un-rangeable, because
```

Comment:
We might not be able to put `s0` in a range, but `i0 <= s0` still can refine `i0` value ranges. Suppose `i0: [a, b]` and `s0: [c, d]`. `i0 <= s0` refinement results will be: `i0: [a, min(b, d)]`.

Reply:
Yeah, there might be some simple stuff possible. In the split example I was looking at, this wouldn't help though, because I typically don't have a range on s0.